src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs

```diff
@@ -112,60 +115,101 @@ public virtual async Task<ProjectPackages> GetInstalledAndTransitivePackagesAsyn
                 return new ProjectPackages(Array.Empty<PackageReference>(), Array.Empty<TransitivePackageReference>());
             }
 
-            IList<LockFileTarget> targetsList = null;
+            IReadOnlyList<LockFileTarget> targetsList = null;
+            T installedPackages;
+            T transitivePackages;
             if (IsInstalledAndTransitiveComputationNeeded)
             {
                 // clear the transitive packages cache, since we don't know when a dependency has been removed
-                InstalledPackages.Clear();
-                TransitivePackages.Clear();
-                targetsList = await GetTargetsListAsync(assetsPath, token);
+                installedPackages = new T();
+                transitivePackages = new T();
+                targetsList = (await GetTargetsListAsync(assetsPath, token))?.ToList();
+            }
+            else
+            {
+                if (InstalledPackages == null || TransitivePackages == null)
+                {
+                    installedPackages = new T();
+                    transitivePackages = new T();
+                }
+                else
+                {
+                    // Don't mutate cache for threadsafety, instead we works on copy then replace cache when done.
+                    lock (_lockObj)
+                    {
+                        (installedPackages, transitivePackages) = GetCacheCopy();
+                    }
+                }
             }
 
             var frameworkSorter = new NuGetFrameworkSorter();
-
             // get installed packages
-            List<PackageReference> installedPackages = packageSpec
+            List<(IReadOnlyList<PackageReference> packageReferences, FrameworkInstalledPackages detectedInstalledPackageChanges)> fetchedInstalledPackagesList = packageSpec
                 .TargetFrameworks
-                .SelectMany(f => FetchInstalledPackagesList(f.Dependencies, f.FrameworkName, targetsList))
+                .Select(f => FetchInstalledPackagesList(f.Dependencies, f.FrameworkName, targetsList, installedPackages)).ToList();
```

Comment:
This implementation made me question where the caching is *actually* happening, and I just realized that we're doing recomputation *even* when everything is up to date. 

There *should* be a spot somewhere in the code that basically just returns the *cached* copies, no computation required. 

Am I missing something? 
dev seems to have the same problem, so this is not behavior introduced in this PR.